chore(coderd/database): enforce agent name unique within workspace build

DanielleMaywood opened the pull request:
This PR creates a database trigger that runs on insert and update of the `workspace_agents` table. The trigger ensures that the agent name is unique within the context of the workspace build it is being inserted into.

DanielleMaywood committed:
chore(coderd/database): create db trigger to enforce agent name uniqueness

This PR creates a new database trigger to ensure an inserted workspace
agent has a unique name within the scope of its workspace.

DanielleMaywood committed:
fix: unique per build not workspace

DanielleMaywood committed:
fix: unique per resource not build

DanielleMaywood committed:
fix: for child agents, ensure unique across provisioner job

Child agents are a new concept so there are very little existing tests
for them. This means we can easily make the change for them without
having to update a mountain of tests. So they get a special case for
checking properly, whilst parent agents have less scrutiny due to the
vast number of tests that appear to not work properly.

DanielleMaywood committed:
chore: fix migration number

deansheather commented on coderd/database/migrations/000331_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,45 @@
+BEGIN
+	IF NEW.parent_id IS NULL THEN
+		-- Count how many agents in this resource already have
+		-- the given agent ID.
```
```suggestion
		-- the given agent name.
```

DanielleMaywood commented on coderd/database/migrations/000331_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,45 @@
+    provisioner_job_id uuid;
+    agents_with_name integer;
+BEGIN
+	IF NEW.parent_id IS NULL THEN
+		-- Count how many agents in this resource already have
+		-- the given agent ID.
+		SELECT COUNT(*) INTO agents_with_name
+		FROM workspace_agents
+		WHERE workspace_agents.resource_id = NEW.resource_id
+		  AND workspace_agents.name = NEW.name
+		  AND workspace_agents.id != NEW.id;
+	ELSE
```
I've explained a bit in the slack thread where we discussed this but attempting to do this for provisioner job instead of resource results in a _lot_ of tests failed.

DanielleMaywood replied:
I've refactored the query a bit and managed to reduce the number of failing tests. I've pushed my changes and you can see the tests that are causing the issue.

One of the guilty tests https://github.com/coder/coder/blob/a18eb9d08fce481afa0a92352a7404090f13ebbe/coderd/insights_test.go#L1321

DanielleMaywood committed:
chore: ID -> name

DanielleMaywood committed:
fix: listen to dean

The CI will fail for this as there are test failures for insights.

DanielleMaywood commented on coderd/database/migrations/000331_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,50 @@
+BEGIN
+	-- Find the provisioner job and workspace the agent is
+	-- being inserted into.
+	SELECT INTO provisioner_job_id, workspace_id_var
```
Can do. Unfortunately we cannot use `workspace_id` as Postgres complains that the use of it in one of the queries is ambiguous.

DanielleMaywood commented on coderd/database/migrations/000331_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,50 @@
+	FROM workspace_resources
+	JOIN provisioner_jobs ON provisioner_jobs.id = workspace_resources.job_id
+	JOIN workspace_builds ON workspace_builds.job_id = provisioner_jobs.id
+	JOIN workspaces ON workspaces.id = workspace_builds.workspace_id
```
At least from my testing, ensuring there is a valid workspace resulted in less test-failures being introduced by this change.

DanielleMaywood commented on coderd/database/migrations/000331_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,50 @@
+
+	-- Count how many agents in this provisioner job already have
+	-- the given agent name.
+	SELECT COUNT(*) INTO agents_with_name
```
Yep. The first is for when you're selecting into a single variable, and the second is for when you're selecting into multiple variables.

DanielleMaywood committed:
fix: use unique name per agent

DanielleMaywood committed:
Merge branch 'main' into dm-workspace-agent-name-uniqueness

DanielleMaywood committed:
chore: feedback

DanielleMaywood committed:
chore: some changes

- replace spaces with tabs.
- remove _var suffixes.
- check if provisioner job as a workspace build

DanielleMaywood committed:
chore: improve sql query

mafredri commented on coderd/database/migrations/000332_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,40 @@
+	JOIN workspace_resources ON workspace_resources.id = workspace_agents.resource_id
+	JOIN workspace_builds ON workspace_builds.job_id = workspace_resources.job_id
+	WHERE workspace_builds.id = workspace_build_id
+	  AND workspace_agents.name = NEW.name
+	  AND workspace_agents.id != NEW.id;
```
```suggestion
		AND workspace_agents.name = NEW.name
		AND workspace_agents.id != NEW.id;
```
Sneaky spaces 😏

mafredri commented on coderd/database/migrations/000332_workspace_agent_name_unique_trigger.up.sql:
```diff
@@ -0,0 +1,40 @@
+CREATE TRIGGER workspace_agent_name_unique_trigger
+	BEFORE INSERT OR UPDATE OF name, resource_id ON workspace_agents
+	FOR EACH ROW
+	EXECUTE FUNCTION check_workspace_agent_name_unique();
```
Maybe we can add a comment why we implement this trigger rather than a unique constraint (e.g. legacy data, etc).

Ideally a `COMMENT ON` if that works for triggers, otherwise plain SQL comment.

mafredri reviewed (approved):
Minor suggestions but otherwise looking good, nice job!

DanielleMaywood committed:
chore: feedback

- add a comment on trigger
- spaces to tabs
